### PR TITLE
Cancel transaction when solution submission time expires

### LINKED ISSUE #821: Solution from previous batch clog solutions for next batch
> ERROR (pod: dev-dfusion-standard-solver-mainnet-master-68bcc79797-2g5r8):
Apr 30 10:05:56.824 ERRO [driver::driver::scheduler::system] Batch 5294136 failed with unretryable error: method 'submitSolution(uint32,uint256,address[],uint16[],uint128[],uint128[],uint16[]):(uint256)' failure: web3 error: RPC error: Error { code: ServerError(-32010), message: "Transaction gas price supplied is too low. There is another transaction with same nonce in the queue. Try increasing the gas price or incrementing the nonce.", data: None }

I am not (yet) sure what this means. I suspect it could be related to the gas price increases we recently implemented.

### LINKED ISSUE #825: [GasEstimation] Don't double gas price if system time is past cutoff time for accepted solutions
We currently double the gas price up to a maximum of 60 GWEI every two blocks regardless of what the current time is. 
In case we already passed the four minute mark up to which solutions are accepted we should probably not double the gas price but instead try to cancel the tx by sending a no-op tx with a slighty higher gas price to clear the tx queue.

https://github.com/gnosis/dex-services/blob/6f794300512e6332e1a80b4ad3ad6700f89fccf1/driver/src/solution_submission/mod.rs#L178-L190

### PR DESCRIPTION
**Pull request by e00E**
There is no reason to continue to try submitting the solution with
higher gas prices when the deadline is already over.
With this commit we cancel the transaction and exit when this happens.

Fixes #825 
Fixes #821 

### Test Plan
Added unit test for this. Wasn't able to test the full behavior with ganache but I did test that sending the noop transaction works in ganache.

### CODE REVIEW

File: core/src/contracts/stablex_contract.rs [hunk 1]
```diff
@@ -324,6 +323,36 @@ impl StableXContract for StableXContractImpl {
...
+                .compat()
+                .await
+                .context("failed to get nonce")?;
+            let transaction_result = ethcontract::contract::MethodBuilder::fallback(
+                web3.clone(),
+                address,
+                Default::default(),
+            )
+            .from(account)
+            .gas(U256::from(21000))
```
Comment (e00E) [inline core/src/contracts/stablex_contract.rs:345]: Is this the right gas amount?

File: core/src/solution_submission.rs [hunk 2]
```diff
@@ -9,10 +9,9 @@ use ethcontract::errors::{ExecutionError, MethodError};
 use ethcontract::web3::types::TransactionReceipt;
 use ethcontract::U256;
 use futures::future::{BoxFuture, FutureExt as _};
-use log::info;
```
Comment (e00E) [inline core/src/solution_submission.rs:12]: This file had mixed uses of logging macros `log::info!` vs `info!`. I prefer having the module name to see where this is from especially for `log::error!`.

File: core/src/contracts/stablex_contract.rs [hunk 3]
```diff
@@ -324,6 +323,33 @@ impl StableXContract for StableXContractImpl {
...
+            let nonce = web3
+                .eth()
+                .transaction_count(address, None)
+                .compat()
+                .await
+                .context("failed to get nonce")?;
+            let transaction_result =
+                ethcontract::contract::MethodBuilder::fallback(web3, address, Default::default())
+                    .from(account)
+                    .gas(U256::from(21000))
```
Comment (e00E) [inline core/src/contracts/stablex_contract.rs:342]: Is this the right gas amount?

File: core/src/contracts/stablex_contract.rs (hunk 1 above)
Comment (nlordell) [reply core/src/contracts/stablex_contract.rs]: AFAIK that is the base transaction gas amount.

File: core/src/solution_submission.rs (hunk 2 above)
Comment (nlordell) [reply core/src/solution_submission.rs:12]: I also prefer this for the log macros. I think a lot of the `use log::{info, error}` come from the migration from edition 2015 to 2018, where the `#[macro_use] extern crate log;` was removed.

File: core/src/contracts/stablex_contract.rs (hunk 3 above)
Comment (nlordell) [reply core/src/contracts/stablex_contract.rs:342]: AFAIK that is the base transaction gas amount.

File: core/src/solution_submission.rs [hunk 4]
```diff
@@ -215,29 +222,60 @@ async fn retry_with_gas_price_increase(
...
+                        gas_price_cap
+                    } else {
+                        U256::from(
+                            (gas_price.as_u128() as f64 * MIN_GAS_PRICE_INCREASE_FACTOR) as u128,
+                        )
+                    };
+                    log::info!(
+                        "trying to cancel solution submission transaction for batch {}",
+                        batch_index
+                    );
```
Comment (e00E) [inline core/src/solution_submission.rs:255]: More logging than necessary maybe but I'd like to have this until we see this working correctly at least once.

File: core/src/contracts/stablex_contract.rs [hunk 5]
```diff
@@ -324,6 +323,35 @@ impl StableXContract for StableXContractImpl {
...
             .boxed()
     }
+
+    fn noop_transaction<'a>(&'a self, gas_price: U256) -> BoxFuture<'a, Result<TransactionResult>> {
+        use futures::compat::Future01CompatExt as _;
+        async move {
+            let web3 = self.instance.raw_instance().web3();
+            let account = self.account().ok_or_else(|| anyhow!("no account"))?;
+            let address = account.address();
+            let nonce = web3
```
Comment (nlordell) [inline core/src/contracts/stablex_contract.rs:333]: This shouldn't be needed, this is how `ethcontract` determines the nonce.

File: core/src/solution_submission.rs [hunk 6]
```diff
@@ -215,29 +222,60 @@ async fn retry_with_gas_price_increase(
...
+                    // `contract.noop_transaction` gets the current nonce from the node which is
+                    // technically a race condition. It would be better to retrieve the nonce from
+                    // the `submit_solution` transaction but I couldn't find a way to do that.
```
Comment (nlordell) [inline core/src/solution_submission.rs:258]: This can be done a couple of ways:
1. Retrieve the transaction data with the transaction hash: https://docs.rs/web3/0.11.0/web3/api/struct.Eth.html#method.transaction. Unfortunately, the transaction hash is not super easy to determine right now and requires first `TransactionBuilder::build`-ing the transaction and using `web3` to send it. I created gnosis/ethcontract-rs#345 to address the issue.
2. Manually setting the nonce for the submit solution transaction by as you do in the `noop_transaction`, and then returning along side the future.

File: core/src/solution_submission.rs [hunk 7]
```diff
@@ -215,29 +222,60 @@ async fn retry_with_gas_price_increase(
...
         // Breaking condition for our loop
-        gas_price_can_still_increase
-            && matches!(
-                result,
-                Err(MethodError {
-                    inner: ExecutionError::ConfirmTimeout,
-                    ..
-                })
-            )
+        match result {
```
Comment (nlordell) [inline core/src/solution_submission.rs:237]: I find this match statement a little hard to follow and it feels a little deeply nested. Maybe it can be refactored a bit? (PS. I have no good ideas on how to do this).

File: core/src/solution_submission.rs (hunk 6 above)
Comment (e00E) [reply core/src/solution_submission.rs:258]: Let's keep it this way until the ethcontract PR is available.

File: core/src/solution_submission.rs (hunk 4 above)
Comment (nlordell) [reply core/src/solution_submission.rs:255]: I think this logging is fine. Its not like this path should happen often, and when it does its nice to have lots of logs in order to determine the exact sequence that was executed.

If this were a common path that would be hit multiple times, then maybe the extra log messages could pollute the logs a bit, but I don't think its the case here.

File: core/src/solution_submission.rs (hunk 7 above)
Comment (e00E) [reply core/src/solution_submission.rs:237]: Agree. Tried this a bit. I think it ends up changing the whole function. In that case I'd like keep this PR as is and make another one to refactor the function.

Comment (nlordell) [reply core/src/solution_submission.rs:238]: Now that I think about it, maybe the function can be refactored in this PR already? Seeing as it is already changing significantly to include the `noop_transaction` to cancel the solution submission?

Comment (e00E) [reply core/src/solution_submission.rs:238]: For the reviewers and when looking at individual commits later it is cleaner to have this separate since the refactoring will change more things than just adding this functionality.

Comment (fleupold) [reply core/src/solution_submission.rs:238]: I cannot follow the code in this match statement (I wasn't even able to properly understand a much simpler flavor [here](https://github.com/gnosis/dex-services/pull/1070#discussion_r449536863)). 

Would it be possible to refactor before adding logic? Maybe we can have a method that returns an `enum SolutionSubmissionAction` and change the loop into something more straight-forward such as:

```rust
loop {
  match next_action(previous_gas_price, gas_price_estimating) {
     SolutionSubmissionAction::Resubmit(gas_price) => { ... }
     SolutionSubmissionAction::LeaveInMemPool => break
  }
}
```

and then in this PR change `LeaveInMemPool` to `CancelTx`. `next_action` could have some unit tests describing what action we expect based on the previous gas price, current estimate and time passed.

File: core/src/solution_submission.rs [hunk 8]
```diff
@@ -358,6 +397,10 @@ mod tests {
             });
         // There should not be a second call to submit_solution because 90 to 100 is not a large
         // enough gas price increase.
+        contract
```
Comment (fleupold) [inline core/src/solution_submission.rs:400]: This test doesn't really match with the comment above. I'm probably missing something, but don't we have to adjust the time somehow to trigger the noop transaction?

File: core/src/solution_submission.rs (hunk 6 above)
Comment (fleupold) [reply core/src/solution_submission.rs:260]: The same race condition applies for increasing the gas price for solution submission.

File: core/src/solution_submission.rs [hunk 9]
```diff
@@ -217,28 +225,59 @@ async fn retry_with_gas_price_increase(
...
+                } else {
+                    let gas_price = if gas_price_can_still_increase {
+                        gas_price_cap
+                    } else {
+                        U256::from(
+                            (gas_price.as_u128() as f64 * min_gas_price_increase_factor).ceil()
+                                as u128,
+                        )
+                    };
+                    log::info!(
```
Comment (fleupold) [inline core/src/solution_submission.rs:254]: Are we cancelling the tx if the gas price can no longer increase? I think in that case we should hope that the current gas price is high enough to complete before the deadline and only cancel after the deadling.

**Review by fleupold — CHANGES_REQUESTED** [review]
My main concern is that the deadline seems to be the one from the solver (3:30) and not the one from the smart contract (4minutes) and that I think we cancel orders even before the deadline arises in case gas becomes too expensive (instead we should leave it in the mempool until the deadline in that case).

It would also be great if we can refactor the do/match statement beforehand to make reasoning about the changes in this PR easier.

File: core/src/solution_submission.rs (hunk 6 above)
Comment (nlordell) [reply core/src/solution_submission.rs:260]: In general though, I think it would be better to ensure the same nonce that was used by the `submit_solution` transaction is used for the cancellation transaction.

File: core/src/solution_submission.rs (hunk 7 above)
Comment (nlordell) [reply core/src/solution_submission.rs:238]: I would also be in favour of a refactor before the change, as this loop has become a bit unruly and would be nice to split up a bit. I like @fleupold's suggestion of a method that tries to submit a solution and returns an enum that can be used to determine what to do next.

Comment (e00E) [reply core/src/solution_submission.rs:238]: I'll do the refactor first in a separate PR then rebase this PR.

File: core/src/contracts/stablex_contract.rs [hunk 10]
```diff
@@ -324,6 +323,27 @@ impl StableXContract for StableXContractImpl {
             .stream()
             .boxed()
     }
+
+    fn noop_transaction<'a>(&'a self, gas_price: U256) -> BoxFuture<'a, Result<TransactionResult>> {
```
[the remark this answers is not in the document]
Comment (nlordell) [reply core/src/contracts/stablex_contract.rs:327]: I'm not a huge fan of `cancel_pending_transaction`, especially since there are no parameter relating to a transaction. I think once we include gnosis/ethcontract-rs#345 then `fn cancel_pending_transaction(self, tx: TransactionResult)` makes a lot more sense, since we would be able to recover the tx nonce and gas price to send a proper transaction cancellation.

Until then `send_noop_transaction` feels better for me. Although, I like using an action verb like `send` to indicate its actually affecting the EVM.

File: core/src/driver/scheduler.rs [hunk 11]
```diff
@@ -5,13 +5,11 @@ use self::evm::EvmScheduler;
 use self::system::SystemScheduler;
 use crate::contracts::stablex_contract::StableXContract;
 use crate::driver::stablex_driver::StableXDriver;
+use crate::models::batch_id::SOLVING_WINDOW;
 use anyhow::{anyhow, Error, Result};
 use std::str::FromStr;
 use std::time::Duration;
 
-/// The time in a batch where a solution may be submitted.
-const SOLVING_WINDOW: Duration = Duration::from_secs(240);
```
Comment (fleupold) [inline core/src/driver/scheduler.rs:13]: nice find!

File: core/src/solution_submission.rs [hunk 12]
```diff
@@ -229,6 +232,24 @@ async fn retry_with_gas_price_increase(
         )
     };
 
+    let wait_for_deadline = || async {
```
Comment (fleupold) [inline core/src/solution_submission.rs:235]: Very clever!

File: core/src/solution_submission.rs [hunk 13]
```diff
@@ -238,15 +259,40 @@ async fn retry_with_gas_price_increase(
...
+                    gas_price_factor,
+                );
+            }
+            Either::Right(_) => {
+                let gas_price = U256::from(
+                    (gas_price.as_u128() as f64 * MIN_GAS_PRICE_INCREASE_FACTOR).ceil() as u128,
+                );
+                match contract.send_noop_transaction(gas_price).await {
+                    Ok(_) => log::info!(
+                        "cancel solution submission of batch {} because of deadline",
```
Comment (fleupold) [inline core/src/solution_submission.rs:282]: ubernit: cancelled

File: core/src/solution_submission.rs [hunk 14]
```diff
@@ -229,6 +232,24 @@ async fn retry_with_gas_price_increase(
...
+        {
+            // Add some extra time because of desync between real current time and ethereum current
+            // block time.
+            Ok(remaining) => async_std::task::sleep(remaining + Duration::from_secs(30)).await,
+            Err(err) => {
+                log::warn!(
+                    "could not calculate remaining time to solve batch: {:?}",
+                    err
+                );
+                future::pending().await
```
Comment (fleupold) [inline core/src/solution_submission.rs:248]: Should this rather resolve immediately? Could this happen if we increased the gas limit with 15s remaining and so the next time we try to compute the remaining time it is negative?

File: core/src/solution_submission.rs [hunk 15]
```diff
@@ -238,15 +259,40 @@ async fn retry_with_gas_price_increase(
...
+                        batch_index
+                    ),
+                    Err(err) => log::error!(
+                        "failed to cancel solution submission of batch {} after deadline: {:?}",
+                        batch_index,
+                        err
+                    ),
+                }
+                // TODO: What to return? Should we make up a method error? Or we could make this
+                // function return Result<(), Either<MethodError, DeadlineReached>> .
```
Comment (fleupold) [inline core/src/solution_submission.rs:292]: Equivalent to your suggestion, I think the cleanest would be our own SolutionSubmissionError enum which can itself be either MethodError or DeadlineReached

File: core/src/solution_submission.rs [hunk 16]
```diff
@@ -238,15 +259,40 @@ async fn retry_with_gas_price_increase(
         } else {
             Some(BLOCK_TIMEOUT)
         };
-        let result = submit_solution(gas_price, block_timeout).await;
-        if !is_confirm_timeout(&result) || is_last_iteration {
-            return result;
+        let solution_future = submit_solution(gas_price, block_timeout);
+        let deadline_future = wait_for_deadline();
+        futures::pin_mut!(deadline_future);
```
Comment (fleupold) [inline core/src/solution_submission.rs:264]: Why is this needed?

File: core/src/solution_submission.rs [hunk 17]
```diff
@@ -238,15 +259,40 @@ async fn retry_with_gas_price_increase(
...
+        match future::select(solution_future, deadline_future).await {
+            Either::Left((result, _)) => {
+                if !is_confirm_timeout(&result) || is_last_iteration {
+                    return result;
```
Comment (fleupold) [inline core/src/solution_submission.rs:268]: Before, we were cancelling as soon as the gas price reached the max. If I read the code correctly, we are now not cancelling at all if the gas price became too expensive. I think we should still cancel in that case, but only after the deadline.

**Review by fleupold — COMMENTED** [review]
Very clever use of `future::select`!

Could you add a few unit test for the new behaviour (specifically around how gas price becoming too expensive and running over time is handled)? 
I can see how this code can be hard to test but given it has a lot of complex logic I think it's really important.

I would think that having a function that computes the next action could make it slightly simpler to test. If need be we can only check for cancellation after each run of submit solution (since we only wait 2 blocks for confirmation) but I do like your parallel solution.

File: core/src/solution_submission.rs (hunk 16 above)
Comment (e00E) [reply core/src/solution_submission.rs:264]: `select` requires the future to be pinned (or unpin). Afaik this is the normal way to accomplish this. It's not needed for `solution_future` because that future happens to already be unpin.

File: core/src/solution_submission.rs (hunk 17 above)
Comment (e00E) [reply core/src/solution_submission.rs:268]: My reasoning is that in the last iteration the solution_future should never return a confirm timeout because for the last iteration we set the block timeout to `None`. So in the last iteration we will either hit the deadline or not get a timeout on the solution.
This isn't clear in the code though so I should add a comment.

File: core/src/solution_submission.rs (hunk 14 above)
Comment (e00E) [reply core/src/solution_submission.rs:248]: Good point.

File: core/src/solution_submission.rs [hunk 18]
```diff
@@ -108,20 +107,31 @@ impl<'a> StableXSolutionSubmitter<'a> {
...
+                            .await
+                        {
+                            return SolutionSubmissionError::from(err);
+                        }
+                    }
                 }
+                SolutionSubmissionError::Unexpected(err.into())
+            }
+            RetryError::TransactionNotConfirmedInTime => {
+                SolutionSubmissionError::Benign("transaction not confirmed in time".to_string())
```
Comment (e00E) [inline core/src/solution_submission.rs:132]: Should this be a benign error?

File: core/src/solution_submission.rs [hunk 19]
```diff
@@ -244,36 +272,86 @@ async fn retry_with_gas_price_increase(
...
+    IncreaseGasPrice,
+    Cancel,
+    Done(std::result::Result<(), RetryError>),
+}
+
+// Either::Left is the solution submission future, Either::Right the deadline future.
+fn determine_retry_action<T0, T1>(
+    either: Either<(Result<(), MethodError>, T0), ((), T1)>,
+    is_last_iteration: bool,
+) -> RetryAction {
```
Comment (e00E) [inline core/src/solution_submission.rs:342]: I know you suggested a test for this but with how simple the function is a test would just repeat the definition so I don't think it's needed.

File: core/src/solution_submission.rs (hunk 18 above)
Comment (fleupold) [reply core/src/solution_submission.rs:132]: I would say this is a bad error, since in this case we are not guaranteed that some solver provided a solution. Thus we should log this as a failure and work towards getting those cases to 0 (we can consider not alerting on them if they are frequent, but they should be measured somewhere).

File: core/src/solution_submission.rs (hunk 16 above)
Comment (fleupold) [reply core/src/solution_submission.rs:304]: I tried to figure out why exactly the deadline future is not unpin, but wasn't able to (might have something to do with `pin_project!` wrapper on the Timeout struct). Do you have an explanation?

File: core/src/solution_submission.rs [hunk 20]
```diff
@@ -244,36 +272,86 @@ async fn retry_with_gas_price_increase(
...
+    match either {
+        Either::Left((result, _)) => {
+            if !is_confirm_timeout(&result) || is_last_iteration {
```
Comment (fleupold) [inline core/src/solution_submission.rs:345]: Do we even need the `|| is_last_iteration` if in the last iteration the confirm will never timeout?

File: core/src/solution_submission.rs (hunk 19 above)
Comment (fleupold) [reply core/src/solution_submission.rs:342]: Agree that just the `determine_retry_action` is not worth testing. I would however like the overall block given how complex it is unless we can factor the timeout in a more orthogonal way (see PR level comment).

We could expose the deadline future part of a trait "wait_for_solution_submission_end" which we can inject and control in unit tests. This would have the added benefit that we could implement it based on the current block timestamp instead of adding 30s to the system time (although I do think that approach is pragmatic and fine too)

**Review by fleupold — COMMENTED** [review]
While investigating why the deadline_future is not unpin I stumbled across this: https://docs.rs/async-std/1.6.2/async_std/future/fn.timeout.html which could be a simpler way to build an "orthogonal" abstraction around this logic. 

We could basically keep the existing `retry_with_gas_price_increase` as the inner future (no logic change needed) and await the Timeout future in a `retry_with_gas_price_increase_and_timout` function.

File: core/src/solution_submission.rs (hunk 16 above)
Comment (e00E) [reply core/src/solution_submission.rs:304]: With medium confidence I believe that that async functions simply do not propagate Unpin. The return type of `wait_for_deadline` is `impl Future` so even if we did nothing in its body it would still be treated as Unpin. The reason that `solution_future` is unpin is because it isn't an `impl Future` but a `BoxFuture`.

File: core/src/solution_submission.rs (hunk 20 above)
Comment (e00E) [reply core/src/solution_submission.rs]: Updated this if condition

File: core/src/solution_submission.rs (hunk 16 above)
Comment (nlordell) [reply core/src/solution_submission.rs:304]: > With medium confidence I believe that that async functions simply do not propagate Unpin

Correct, `async` functions are marked as `!Unpin` regardless of their body. For example even the following `async` function is `!Unpin`:
```rust
async fn answer() -> i32 { 42 }
```
In fact, an `async` functions get wrapped by the `const fn std::future::from_generator<T>(gen: T) -> impl Future<Output = T::Return>`, and their underlying type is explicitly `!Unpin`: 
https://github.com/rust-lang/rust/blob/8aa18cbdc5d4bc33bd61e2d9a4b643d87f5d21de/src/libcore/future/mod.rs#L64

> is unpin is because it isn't an impl Future but a BoxFuture

You could also `.boxed()` instead of `futures::pin_mut!`. You might also be able to `.fuse()` the future as well which I thought was needed for `select!`.

File: core/src/solution_submission.rs [hunk 21]
```diff
@@ -174,77 +184,175 @@ impl<'a> StableXSolutionSubmitting for StableXSolutionSubmitter<'a> {
     }
```
Comment (nlordell) [inline core/src/solution_submission.rs:184]: I think the easier solution would be to:
```rust
let mut submission = retry_with_gas_price_increase(...);
let mut deadline = wait_for_deadline(...);

select! {
    result = submission => result.map_err(self.make_error(...)),
    _ = b => {
        self.contract.send_noop_transaction(...);
        Err(...)
    },
};
```

**Review by nlordell — CHANGES_REQUESTED** [review]
I don't think we need to select on the deadline inside the retry method but instead around it. I think this is what @fleupold mentioned.

File: core/src/solution_submission.rs (hunk 21 above)
Comment (e00E) [reply core/src/solution_submission.rs:184]: I thought about this. The disadvantage is that we don't know what gas price was used so we have to be maximally pessimistic. Since this hopefully happens rarely it is probably worth it for the simpler code. Will try and see how it looks.

**Comment by e00E** [comment]
This is ready. I implemented the suggestion for handling the deadline one level higher and it turns out much cleaner. We still can't easily test that level but imo it is sufficiently simple that it shouldn't be a problem.

File: core/src/solution_submission.rs [hunk 22]
```diff
@@ -155,19 +159,48 @@ impl<'a> StableXSolutionSubmitting for StableXSolutionSubmitter<'a> {
         solution: Solution,
         claimed_objective_value: U256,
     ) -> BoxFuture<Result<(), SolutionSubmissionError>> {
+        const GAS_PRICE_CAP: u64 = 60_000_000_000;
         async move {
-            match retry_with_gas_price_increase(
+            let submit_future = retry_with_gas_price_increase(
```
Comment (nlordell) [inline core/src/solution_submission.rs:164]: I think `submit_solution` method should take a `nonce` parameter, and we should use `web3` here to read it. That way the `send_noop_transaction` is guaranteed to have the same nonce as the solution submission transaction.

I would be OK with adding that in a follow up PR though.

**Review by nlordell — APPROVED** [review]
Looks good. Just one suggestions (but I would be OK in including it as a follow up PR).

**Comment by nlordell** [comment]
>  We still can't easily test that level but imo it is sufficiently simple that it shouldn't be a problem.

Maybe we should mock a `Deadline` trait that creates a deadline future to unit test the error condition?

**Review by fleupold — APPROVED** [review]
Looks very nice

> Maybe we should mock a Deadline trait that creates a deadline future to unit test the error condition?

I agree that testing this would be nice and should be possible by not depending directly on async_std::future::timeout in the code. Given the rest of the logic is fairly well covered in the tests I think we should add coverage for this part as well.

File: core/src/models/batch_id.rs [hunk 23]
```diff
@@ -46,6 +48,10 @@ impl BatchId {
         self.order_collection_start_time() + BATCH_DURATION
     }
 
+    pub fn solve_end_time(self) -> SystemTime {
+        self.solve_start_time() + SOLVING_WINDOW
+    }
```
Comment (bh2smith) [inline core/src/models/batch_id.rs:53]: What happens if `self.solve_start_time() + SOLVING_WINDOW` is greater than the batch length? Do we have anything checking or imposing such a constraint?

**Review by bh2smith — APPROVED** [review]
Looks good, just one concern about the solve stop time.

File: core/src/models/batch_id.rs (hunk 23 above)
Comment (e00E) [reply core/src/models/batch_id.rs:53]: The constants for batch duration and solving window are defined right next to each other with one clearly being larger than the other so I don't think there is a need to assert this more.

Comment (nlordell) [reply core/src/models/batch_id.rs:53]: Rust needs `static_assert!`s :stuck_out_tongue:
